app/gui/view/graph-editor/src/lib.rs

```diff
@@ -1681,6 +1683,99 @@ impl GraphEditorModel {
         find_free_place(starting_from, direction, node_areas)
     }
 
+    pub fn new_node_position_after_dropping_edge(
+        &self,
+        edge_id: EdgeId,
+        mouse_position: Vector2,
+    ) -> Vector2 {
+        let edge_source_node_id = self.edge_source_node_id(edge_id);
+        let edge_source_node = edge_source_node_id.and_then(|id| self.nodes.get_cloned_ref(&id));
+        self.new_node_position_based_on_mouse(mouse_position, edge_source_node)
+    }
+
+    pub fn new_node_position_based_on_mouse_without_source(
+        &self,
+        mouse_position: Vector2,
+    ) -> Vector2 {
+        let nearest_node = self.find_nearest_node(mouse_position);
+        self.new_node_position_based_on_mouse(mouse_position, nearest_node)
+    }
+
+    fn find_nearest_node(&self, mouse_position: Vector2) -> Option<Node> {
+        let mut min_distance_squared = f32::MAX;
+        let mut nearest_node = None;
+        let nodes = self.nodes.all.raw.borrow();
+        for node in nodes.values() {
+            // For calculating the approximate distance between a node and the mouse pointer, one
+            // of several possible reference points on a node can be chosen. In order of increasing
+            // complexity and precision:
+            //  a) node.position(),
+            //  b) the geometrical center of the node's area,
+            //  c) a point closest to the mouse pointer on the line segment connecting node's
+            //     leftmost & rightmost points,
+            //  d) a point closest to the mouse pointer on the bounding box of the node
+            //     (with extra handling for when the mouse is inside the bounding box),
+            //  e) a point closest to the mouse pointer on the rounded rectangle forming the node's
+            //     shape (with extra handling for when the mouse is inside the rounded rectangle).
+            //  The following code implements variant (c), the first of the listed ones that
+            //  resulted in a UX which feels reasonably predictable and intuitive.
+            let node_position = node.position().xy();
+            let node_rightmost_x = node_position.x + node.model.width();
+            let mouse_x = mouse_position.x;
+            let node_point_near_mouse = if mouse_x <= node_position.x {
+                node_position
+            } else if mouse_x < node_rightmost_x {
+                Vector2(mouse_x, node_position.y)
+            } else {
+                Vector2(node_rightmost_x, node_position.y)
+            };
+            let distance_squared = (node_point_near_mouse - mouse_position).norm_squared();
+            if distance_squared < min_distance_squared {
+                min_distance_squared = distance_squared;
+                nearest_node = Some(node.clone_ref());
+            }
+        }
+        nearest_node
+    }
+
+    pub fn new_node_position_based_on_mouse(
```

Comment:
If you go to any other developer and ask them "what does the function `new_node_position_based_on_mouse` do?", would they answer what it really does? After reading the name I do not know. Maybe if it had documentation it owuld be easier. I'm skipping checking the code of the function until it has docs or understandable name. However, please be sure that all the variables in this and other functions are clear and understandable. I did a fast glance at its body and I do not understand what most of the names mean there. Either find better names, or write a proper documentation for it – reading the documentation should give the reader **the same information as the design doc** – does right now the documentation cover **all** the design doc elements?